server/src/main/java/org/elasticsearch/action/admin/indices/refresh/TransportShardRefreshAction.java

```diff
@@ -73,21 +78,30 @@ protected ReplicationResponse newResponseInstance(StreamInput in) throws IOExcep
     protected void shardOperationOnPrimary(
         BasicReplicationRequest shardRequest,
         IndexShard primary,
-        ActionListener<PrimaryResult<BasicReplicationRequest, ReplicationResponse>> listener
+        ActionListener<PrimaryResult<ReplicaShardRefreshRequest, ReplicationResponse>> listener
     ) {
         ActionListener.completeWith(listener, () -> {
-            primary.refresh(SOURCE_API);
+            var refreshResult = primary.refresh(SOURCE_API);
             logger.trace("{} refresh request executed on primary", primary.shardId());
-            return new PrimaryResult<>(shardRequest, new ReplicationResponse());
+            var shardRefreshRequest = new ReplicaShardRefreshRequest(primary.shardId(), refreshResult.generation());
+            // Since we are not reusing the same request on the replica shards and create a new ReplicaShardRefreshRequest, we
+            // need to explicitly set the parent task so the resulting refresh[s][r] task has the same refresh[s] task as parent.
```

Comment:
I know this was a hard-won lesson but I don't think it's best to capture it in a comment like this. Could we instead capture this requirement in an assertion, for example in the following location? That way, if anyone else encounters it in a different situation they will get a nice actionable failure straight away?

https://github.com/elastic/elasticsearch/blob/43b69b6f704da5b0a41f1785635f2c894bd9d2e5/server/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java#L1143-L1144

Reply:
Do you mean just asserting that the parent task id is set? In our case the parent of the `ReplicaRequest` and the parent of the `Request` must be the same, but I'm not sure we can assert that is always the case.

Reply:
I'd try the stronger invariant first: the parent IDs should match. If we find any cases where that isn't true, we can think about how to address them.